Be more strict with Composer version constraints

fredden opened the pull request:
This pull request changes the Composer version constraint normalisation to be more opinionated / strict. 

Please let me know if you would prefer to have these enhancements as individual / separate pull requests. It may be that some of these rules are desired whereas others are too subjective / opinionated for now.

Changes include:

- ~*and- constraints are always separated by a single space (` `), never a comma (`,`)~ https://github.com/ergebnis/json-normalizer/pull/819
- ~overlapping constraints are removed (`^1.0 || ^1.1 || ^2.0` -> `^1.0 || ^2.0`)~ https://github.com/ergebnis/json-normalizer/pull/850
- tilde operators are preferred over wildcard version ranges (`1.0.*` -> `~1.0.0`)
- caret operators are preferred over tilde operators (`~1` -> `^1`, `~1.3` -> `^1.3`)
- version numbers have a minimum number of parts (`^1` -> `^1.0`)
- patch number is trimmed when zero for caret operator (`^1.0.0` -> `^1.0`)
- ~version numbers are sorted in ascending order (`^2.0 || ^1.4` -> `^1.4 || ^2.0`)~ https://github.com/ergebnis/json-normalizer/pull/816

fredden committed:
Be more strict with Composer version constraints

fredden committed:
Move composer/semver to 'require' from require-dev

fredden committed:
Improve test coverage (to appease mutation check)

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Handle inline aliases

fredden committed:
Remove rogue trailing 'or'

fredden committed:
Update template files to match expected results

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

fredden committed:
Upgrade minimum version of composer/semver

Tests fail with version 3.2.0, but work fine with 3.2.1+

fredden committed:
Increase test coverage

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Fix: Widen version constraint

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Fix: Revert

localheinz commented on test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json:
```diff
@@ -1,11 +1,11 @@
-        "combination-or-version-range-caret-not-overlapping/04-with-single-space-untrimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/05-with-double-space-trimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/06-with-double-space-untrimmed": "^0 || ^1.2 || ^2.3.4"
+        "combination-or-version-range-caret-not-overlapping/01-without-spaces-trimmed": "^0.0 || ^1.2 || ^2.3.4",
```
@fredden 

I prefer to always use `major.minor.patch` versions, how about you?

What do you think about normalizing

- `major`
- `major.minor`

to `major.minor.patch` versions?

fredden replied:
Given a choice of `^1`, `^1.0`, `^1.0.0`, I prefer `^1.0`.

For me, this strikes a good balance between having too much and too little information. Typically the bugfix/patch version isn't relevant for a package requirement when semver is in play, as the developer specifying the dependency only usually cares about a specific feature being present (so major.minor) and any patch releases are a bonus. When the patch version is relevant (see recent commits in this branch for `composer/semver`), it's never `.0` that's specified.
So, when I see `^1.2.3`, I know that 1.2.0 has a bug that's relevant for the correct functioning of the package that's asking for it. But `^6.3.0` is equivalent to `^6.3` and doesn't communicate any extra information for me, so can be omitted. However `^22` feels like it's too vague to me.

I think we should make a decision and then always normalise to this format. Implementing either format will be a single line of code, which I'm happy to add when you have decided which way to go.

fredden replied:
Another thought on this: when I run `composer require squizlabs/php_codesniffer` (with no version constraint specified), Composer adds `    "squizlabs/php_codesniffer": "^3.7",` to `composer.json` (and installs version `3.7.1` in `composer.lock` -- which is the latest stable version at time of writing).

localheinz committed:
Enhancement: Add examples

localheinz committed:
Fix: Indentation

localheinz committed:
Fix: Indentation

localheinz committed:
Fix: Remove

localheinz committed:
Fix: Remove unnecessary statement

localheinz committed:
Fix: Use boolean expression

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Trim patch number when zero for caret operator

Some discussion about this is underway in the pull request and on social media.
It may be decided to always specify the patch number (even when zero) instead
of trimming when zero. If that decision is made, we can revert this and
implement that normalisation instead.

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Remove overlapping constraints

fredden committed:
Increase test coverage

fredden committed:
Increase test coverage

fredden committed:
Move new tests to correct files

localheinz committed:
Enhancement: Extract method

localheinz committed:
Enhancement: Update CHANGELOG.md

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'origin/remove-overlapping' into version-normalisation

fredden committed:
Correct whitespace with php-cs-fixer

fredden committed:
Remove stray changelog entry

I missed this when merging another branch into this. Eventually this branch will need a changelog entry too.

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

fredden committed:
Merge remote-tracking branch 'upstream/main' into version-normalisation

localheinz committed:
Merge branch 'main' into version-normalisation

localheinz committed:
Fix: Remove adjustment for parts (for now)

localheinz committed:
Fix: Update CHANGELOG.md

localheinz reviewed (approved):
👍
